### ISSUE TITLE
期待select option支持number/integer类型，否则真的太痛苦了

### ISSUE THREAD

**Issue by eclanp** [ant-design/ant-design#2857]
一般关联的外键都是主键ID，都为integer类型，现在option的value只能是string类型，感觉用起来很不舒服啊，这个可以优化么？

**Comment by codering** [comment]
同感 https://github.com/react-component/select/issues/115

**Comment by whx76** [comment]
同感，包括 treeselec也有这个问题，要来回转换

**Comment by infeng** [comment]
同感

**Comment by batouy** [comment]
是的，之前写的校验规则，设置type为number，value都是用的number类型，现在都得改了，忧伤啊

**Comment by whx76** [comment]
@batouy 我也是这样改了两遍

**Comment by slertness** [comment]
确实 现在radio的value还是可以用number的 感觉是一种不一致

**Comment by whx76** [comment]
期待这个能有调整 支持int类型

**Comment by yesmeck** [comment]
@afc163 打算做调整吗？调整的话我发个 PR。

**Comment by afc163** [comment]
这个要问 @yiminghe

**Comment by whx76** [comment]
期待调整啊

**Comment by afc163** [comment]
trace in https://github.com/react-component/select/pull/120

**Comment by dreamcatcher081210** [comment]
看到这个issue我眼泪流出来了，原来这不是矫情，是同病相怜，而对于toString()，我已经吐了

**Comment by starInEcust** [comment]
为啥升到2.7.1还是有这个问题。。。还木有搞定吗

**Comment by afc163** [comment]
@starInEcust 

1. This issue is still open, so it is still a unresolved problem.
2. I believe there is not any changelog about Select'value for now.

**Comment by vxzhong** [comment]
+1, really need this. A lot of warning happened because of this issue.
Why https://github.com/react-component/select/pull/120 is closed, but it's not merged?

**Comment by whx76** [comment]
treeselec also

**Comment by bytedaring** [comment]
怎么解决的？

**Comment by yvanwangl** [comment]
期待+1

**Comment by cnfi** [comment]
期待+1

**Comment by slertness** [comment]
期待+1

**Comment by afc163** [comment]
@yesmeck number 应该支持起来比较容易，其他更多类型就不支持了。

**Comment by cc258** [comment]
2.12.4 mark一下

**Comment by RequiemSoul** [comment]
也碰到了这个问题  希望能解决

**Comment by davebream** [comment]
Looking forward +1

**Comment by clodal** [comment]
Is anyone working on a PR for this issue? Seems pretty long-standing, I could give it a go if it's not being worked on

**Comment by 282159468** [comment]
++++1

**Comment by callmemb** [comment]
Why only number ?  I would love to use objects, booleans and other in option.

**Comment by wangmeijian** [comment]
还没解决？Is that settled?

**Comment by yesmeck** [comment]
antd 3 会支持

**Comment by fubobing126** [comment]
select 多选 限制个数这个加上去了吗

**Comment by yesmeck** [comment]
Supported in antd 3.

**Comment by yejinjian** [comment]
用了 antd 3 并没有支持丫, 哭晕在厕所

**Comment by yesmeck** [comment]
@yejinjian 如果确认有问题，新建一个 issue 吧。

**Comment by hjin-me** [comment]
3.0 支持 number 了么？测试好像失败了。文档里面也没有说明。

**Comment by yesmeck** [comment]
@hjin-me 有问题新建 issue，带上重现 demo。

**Comment by jagratiTomar** [comment]
any update on the issue

**Comment by adamchenwei** [comment]
did the change been rolled back? its 3.9 + already, but still, TreeSelect is still not allowing array of number.... is there a reason for it?
https://codesandbox.io/s/7wlrk3q9xq

**Comment by storesbuzz** [comment]
AutoComplete Options are also not supporting value as integer. getting error - 

Uncaught Error: Invalid `value` of type `number` supplied to Option, expected `string` when `tags/combobox` is `true`.

kindly support here, values are most commonly used as intergers only.

**Comment by zch233** [comment]
> Failed prop type: Invalid prop `value` supplied to `Option`.

希望能支持多类型数据，如 boolean、null。

**Comment by Jijmin** [comment]
Is the same problem solved? Using `typeof` is already a `string`, but still gives an error.

"antd": "^3.26.7",
"react": "^16.12.0",
"node": "v10.16.3"

```js
getOptionList(data) {
  if (!data) {
    return [];
  }
  let options = []; //[<Option value="0" key="all_key">全部</Option>];
  data.map(item => {
    console.log(item.id.toString(), typeof item.id.toString()) // 0 string // 1 string // 2 string // ...
    options.push(
      <Option value={item.id.toString()} key={item.id}>
        {item.name}
      </Option>
    );
  });
  return options;
}

const data = [
  { id: "0", name: "全部" },
  { id: "1", name: "北京" },
  { id: "2", name: "天津" },
  { id: "3", name: "上海" }
]

<Select>
  {getOptionList(data)}
</Select>
```
```
index.js:1 Warning: Failed prop type: Invalid prop `id` of type `array` supplied to `Select`, expected `string`.
    in Select (at baseForm/index.js:64)
    in FilterForm (created by Form(FilterForm))
    in Form(FilterForm) (at order/index.js:94)
    in div (created by Context.Consumer)
    in div (created by Context.Consumer)
    in Card (at order/index.js:93)
    in div (at order/index.js:92)
    in Order (created by Context.Consumer)
    in Route (at router.js:52)
    in Switch (at router.js:36)
    in div (created by Context.Consumer)
    in Row (at Main.js:17)
    in div (created by Context.Consumer)
    in Col (at Main.js:15)
    in div (created by Context.Consumer)
    in Row (at Main.js:11)
    in Admin (at router.js:35)
    in Route (at router.js:32)
    in Switch (at router.js:30)
    in div (at App.js:5)
    in App (at router.js:29)
    in Router (created by HashRouter)
    in HashRouter (at router.js:28)
    in ERouter (at src/index.js:7)
```

**Comment by hikaruchang** [comment]
You can use value={number} as integer

**Comment by creativiii** [comment]
This is still unresolved for me as well. Did the change get rolled back?

**Comment by jannikbuschke** [comment]
+1

**Comment by balboFK** [comment]
Really need to use a object as a value too. Take a look at how Vuetify handles this, should be a good API example:

https://vuetifyjs.com/en/api/v-select/#props-items

It should accept objects and numbers as a value aswell.

**Comment by mrlubos** [comment]
> Supported in antd 3.

https://user-images.githubusercontent.com/12529395/135748052-6f93e19a-8e9e-4435-a4c4-744e92a4892f.mp4

**Comment by balboFK** [comment]
What is the status on this feature? 
@afc163 @zombieJ 
I'm in need of a object as a value, and this time I dont think I can avoid it.

**Comment by mehran-dev** [comment]
if the value of option cant accept arrayOfObjects it really annoying , and stupid !

**Comment by osamamammar** [comment]
I have the same problem... why this issue is closed? I'm using antd version 4.23.0
is there any solution? I want to add a number for value on the select option

**Comment by JulianNymark** [comment]
My solution (as I've seen others do for this exact case) is to use `JSON.stringify()` when creating Option objects, and then in my custom render functions I must remember to json parse it... it's not the best... in fact it's kinda yucky, but hey it works! you can put _anything_ in a string! 

Just make sure that the original "value" string is reflected in the object that you stringify (just to avoid subtle weirdness later on, this enforces uniqueness of the resultant stringified string, since I believe it's sorted and filtered on in the ant component implementations)

Since example snippets are worth a thousand words:

```typescript
const enrich_tags = (tag_list: string[]): Partial<CustomTagProps>[] => {
    return tag_list.map((element: string) => {
      if (isValidColor(element)) {
        return { label: element, value: JSON.stringify({value: element, color: element}) }
      }
      return { label: element, value: JSON.stringify({value: element, color: ''}) }
    })
  }
```

I just wanted to encode the color of a tag along with its normal string value.

now I have to access `value.value...` not the greatest, but it's ok... I'm not going crazy.
